[ojelectronics] Initial contribution

EvilPingu opened the pull request:
Hi,

this is the initial contribution for ojelectronics binding. First of all i build two things:

- OJCloud: Handles the traffic with the ojelectronics cloud
- Thermostat: Have the (read only) channel for the associated thermostats

OJ Electronics builds floor heating thermostats. Goal of the binding is to support OJ Electronics devices.

EvilPingu

fwolter commented:
Can you fix the conflicting file and the build errors?

EvilPingu commented:
> Can you fix the conflicting file and the build errors? I would take a look into it, then.

Hi,

conflicts are solved now.

EvilPingu

fwolter commented on bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/services/RefreshService.java:
```diff
@@ -0,0 +1,158 @@
+                    refreshDone.accept(content);
+                }
+            } catch (JsonSyntaxException exception) {
+                handleConnectionLosed();
```
You might want to log the exception's message.

fwolter commented on bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/services/SignInService.java:
```diff
@@ -0,0 +1,98 @@
+        });
+    }
+
+    private PostSignInQueryModel getpostSignInQueryModel() {
```
```suggestion
    private PostSignInQueryModel getPostSignInQueryModel() {
```

fwolter commented on bundles/org.openhab.binding.ojelectronics/src/main/resources/ESH-INF/i18n/ojelectronics_de_DE.properties:
```diff
@@ -0,0 +1,18 @@
+# FIXME: please do not add the file to the repo if you add or change no content
+# binding
+binding.ojelectronics.name = "Binding f�r OJElectronics"
+binding.ojelectronics.description = "Bindingf�r OJElectronics."
```
```suggestion
binding.ojelectronics.description = "Binding f�r OJElectronics."
```

fwolter commented on bundles/org.openhab.binding.ojelectronics/src/main/resources/ESH-INF/i18n/ojelectronics_de_DE.properties:
```diff
@@ -0,0 +1,18 @@
+# FIXME: please substitute the xx_XX with a proper locale, ie. de_DE
+# FIXME: please do not add the file to the repo if you add or change no content
```
Can these be removed?

fwolter commented on bundles/org.openhab.binding.ojelectronics/src/main/resources/ESH-INF/i18n/ojelectronics_de_DE.properties:
```diff
@@ -0,0 +1,18 @@
+
+# thing type config description
+thing-type.config.ojelectronics.bridge.userName.label = "Nutzername"
+thing-type.config.ojelectronics.bridge.userName.description = "Nutzername f�r den Zugriff auf OJ Electronics cloud."
```
```suggestion
thing-type.config.ojelectronics.bridge.userName.description = "Nutzername f�r den Zugriff auf OJ Electronics Cloud."
```

fwolter commented on bundles/org.openhab.binding.ojelectronics/src/main/resources/ESH-INF/i18n/ojelectronics_de_DE.properties:
```diff
@@ -0,0 +1,18 @@
+# FIXME: please substitute the xx_XX with a proper locale, ie. de_DE
```
This file is encoded with ISO-8859-1. You might want to use UTF-8.

fwolter commented on bundles/org.openhab.binding.ojelectronics/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,165 @@
+				<advanced>true</advanced>
+				<default>https://OWD5-OJ001-App.ojelectronics.com/api</default>
+			</parameter>
+			<parameter name="refreshDelayInSeconds" type="integer" required="true" min="15">
```
```suggestion
			<parameter name="refreshDelayInSeconds" type="integer" required="true" min="15" unit="s">
```

fwolter commented on bundles/org.openhab.binding.ojelectronics/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,165 @@
+				<default>https://OWD5-OJ001-App.ojelectronics.com/api</default>
+			</parameter>
+			<parameter name="refreshDelayInSeconds" type="integer" required="true" min="15">
+				<label>Refresh delay</label>
```
Words in labels should be capitalized. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

fwolter commented on bundles/org.openhab.binding.ojelectronics/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,165 @@
+				<context>password</context>
+				<default>f219aab4-9ac0-4343-8422-b72203e2fac9</default>
+			</parameter>
+			<parameter name="softwareVersion" type="integer" required="true">
```
Should this be configureable by the user or is it more a constant, that depends on your current implementation?

fwolter commented on bundles/org.openhab.binding.ojelectronics/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,165 @@
+		<item-type>Number</item-type>
+		<label>Floor temperature</label>
+		<category>Temperature</category>
+		<state pattern="%.1f °C" readOnly="true" />
```
If you use UoM
```suggestion
		<state pattern="%.1f %unit%" readOnly="true" />
```

fwolter commented on bundles/org.openhab.binding.ojelectronics/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,165 @@
+		</config-description>
+	</thing-type>
+	<channel-type id="floorTemperature">
+		<item-type>Number</item-type>
```
```suggestion
		<item-type>Number:Temperature</item-type>
```

fwolter commented on bundles/org.openhab.binding.ojelectronics/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,165 @@
+			</options>
+		</state>
+	</channel-type>
+	<channel-type id="serialNumber">
```
Is it necessary to expose the serialNumber as a Channel, since it is already a property.

fwolter commented on bundles/pom.xml:
```diff
@@ -178,6 +178,8 @@
     <module>org.openhab.binding.nuki</module>
     <module>org.openhab.binding.oceanic</module>
+    <module>org.openhab.binding.ojelectronics</module>
+
```
Can you remove the blank line?

fwolter commented on bundles/pom.xml:
```diff
@@ -509,3 +511,4 @@
   </profiles>
 
 </project>
+
```
See above

fwolter commented on bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/OJElectronicsBridgeHandler.java:
```diff
@@ -0,0 +1,177 @@
+package org.openhab.binding.ojelectronics.internal;
+
+import java.util.Map;
+import java.util.concurrent.Executors;
```
This is not used.

fwolter commented on bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/services/SignInService.java:
```diff
@@ -0,0 +1,98 @@
+                    connectionLosed.run();
+                    return;
+                }
+                if (result != null && result.getResponse().getStatus() != 200) {
```
This null check is not needed.

EvilPingu commented on bundles/org.openhab.binding.ojelectronics/README.md:
```diff
@@ -0,0 +1,48 @@
+
+| Thing                | Type   | Description                         |
+|----------------------|--------|-------------------------------------|
+| bridge               | Bridge | OJ Electronics Bridge               |
```
The bridge is a wrapper around the REST-API. It is not a concrete product but every product like the thermostats us it. I think bridge is the right word for it. But i think i could give the bridge a more meaningful name like "API bridge".

EvilPingu commented on bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/OJElectronicsThermostatHandler.java:
```diff
@@ -0,0 +1,124 @@
+     */
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        // nothing do here
```
Yes i want but not an the moment. Handle command is the next step

EvilPingu commented on bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/config/OJElectronicsBridgeConfiguration.java:
```diff
@@ -0,0 +1,59 @@
+    /**
+     * API-Key
+     */
+    public String APIKEY = "f219aab4-9ac0-4343-8422-b72203e2fac9";
```
It's not public. Should i remove the bridge parameter?

fwolter replied:
It depends on the license of the vendor. If you are allowed to distribute the key to third parties, it's ok. By signing the commit messages you take responsibility.

EvilPingu replied:
I think it's complicated. I'm not the vendor so i'm not allowed to distrubute. So i'm deleting the Key and every one have to set in a thing property.

EvilPingu commented on bundles/org.openhab.binding.ojelectronics/README.md:
```diff
@@ -0,0 +1,48 @@
+| groupName         | Text   | Group name                   |
+| groupId           | Number | Group Id                     |
+| online            | Switch | Online                       |
+| heating           | Switch | Heating                      |
```
I think a Contact is for windows and doors, isn't it? Should i change it anyway?

fwolter replied:
`Contact` is an abstract contact that can be open and closed. It doesn't relate to windows and doors at first sight, but can also be a signal input. I think that fits better than a switch.

EvilPingu replied:
ok. I changed it

EvilPingu commented:
> 
> 
> Very professional code. I like the usage of Jetty's async API and the lambdas.
> 
> There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html`.
> 
> There are some formatting issues. You can view them with `mvn spotless:check -Dspotless.check.skip=false` and fix them with `mvn spotless:apply`.

Hi @fwolter,

what are the next steps? How do we get on? Did I miss something?

EvilPingu commented:
> Thanks for the hint. Github doesn't notify when comments are set to resolved.
> 
> I found a few minor things I overlooked during the first review. Sorry for that!
> 
> There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html`.

Hi @fwolter,

i fixed all review things and i also checked target/code-analysis/report.html. It's empty.

@EvilPingu

fwolter commented:
Did you run `mvn install`? I see 15 warnings:
![grafik](https://user-images.githubusercontent.com/1894841/86264823-45a9e380-bbc3-11ea-967a-694929c7c2a6.png)

EvilPingu commented:
> Did you run `mvn install`? I see 15 warnings:
> ![grafik](https://user-images.githubusercontent.com/1894841/86264823-45a9e380-bbc3-11ea-967a-694929c7c2a6.png)

I didn't understand this. I'm confused. I make "mvn clean install". No errors, no warning: file:///home/ckit/dev/openhab-2-5-x/git/openhab-addons/target/code-analysis/report.html and file:///home/ckit/dev/openhab-2-5-x/git/openhab-addons/target/summary_report.html

fwolter commented:
Did you push your changes? For example RefreshService:56, the field name is still lower case.

EvilPingu commented:
It's still lowercase. But why i don't get any warnings?

I' running: mvn -T 11 clean install -pl :org.openhab.binding.ojelectronics

Do I look in the wrong files?

J-N-K commented:
Depending on the time you started the branch, you could still run an old version of SAT. Which version does it show?

EvilPingu commented:
[INFO] --- sat-plugin:0.9.0:checkstyle (sat-all) @ org.openhab.binding.ojelectronics ---
[INFO] Adding dependency to checkstyle:0.9.0
[INFO] Adding dependency to checkstyle:8.12
[WARNING] Could not find file feature file features/karaf/openhab-addons/src/main/feature/feature.xml
[INFO] 
[INFO] --- sat-plugin:0.9.0:pmd (sat-all) @ org.openhab.binding.ojelectronics ---
[INFO] Adding dependency to pmd:0.9.0
[INFO] Adding dependency to pmd-core:6.7.0
[INFO] Adding dependency to pmd-java:6.7.0
[INFO] Adding dependency to pmd-javascript:6.7.0
[INFO] Adding dependency to pmd-jsp:6.7.0
[INFO] 
[INFO] --- sat-plugin:0.9.0:spotbugs (sat-all) @ org.openhab.binding.ojelectronics ---
[INFO] Adding dependency to findbugs:0.9.0
[INFO] Adding dependency to bug-pattern:1.5.0
[INFO] Adding dependency to spotbugs:3.1.12
[INFO] Fork Value is true
[INFO] Done SpotBugs Analysis....
[INFO] 
[INFO] --- sat-plugin:0.9.0:report (sat-all) @ org.openhab.binding.ojelectronics ---
[INFO] Empty report will not be appended to the summary report.

EvilPingu commented:
I also check this one: [https://ci.openhab.org/job/PR-openHAB2-Addons/20216/Static_20Code_20Analysis_20Report/](url)

No problems

J-N-K commented:
Bingo. There was a bug in 0.9.0. Do you know how to rebase on 2.5.x? There should be no conflicts.

EvilPingu commented:
> Bingo. There was a bug in 0.9.0. Do you know how to rebase on 2.5.x? There should be no conflicts.

Not exactly. Can you please try to help me

J-N-K commented:
It‘s a bit difficult since your branch is named 2.5.x. Do you have different remotes configured for the original openhab repo and your fork?

EvilPingu commented:
I know, after the pull request is commited, i want to delete my fork and create a new one with a better git configuration.

This is my git configuration:

ckit@ubuntu-dev2:~/dev/openhab-2-5-x/git/openhab-addons(bugfixing)$ git remote -v
2.5.x	https://github.com/openhab/openhab-addons.git (fetch)
2.5.x	https://github.com/openhab/openhab-addons.git (push)
EvilPingu	https://github.com/EvilPingu/openhab-addons.git (fetch)
EvilPingu	https://github.com/EvilPingu/openhab-addons.git (push)
origin	https://github.com/openhab/openhab-addons.git (fetch)
origin	https://github.com/EvilPingu/openhab-addons.git (push)

Normally i'm working and pushing to  a branch namend "bugfixing". Afterwards i'm pushing to 2.5.x.

Sorry for the trouble. I'm definitely no git expert.

J-N-K commented:
In the base pom.xml the sat Version is defined. It may be easier to change that from 0.9.0 to 0.11.0, fix the errors and revert that change before committing.

EvilPingu commented:
There is only a version 0.10.0.

I will fixe the problem with the lowercase static field. This is the only thing sat has found.

fwolter commented:
This is quite strange, since we use the same code base, we should get the same results. Here is my report. The NewlineAtEndOfFileCheck are false positives and can be ignored. [summary_report.zip](https://github.com/openhab/openhab-addons/files/4860532/summary_report.zip)

EvilPingu commented:
> 
> 
> This is quite strange, since we use the same code base, we should get the same results. Here is my report. The NewlineAtEndOfFileCheck are false positives and can be ignored. [summary_report.zip](https://github.com/openhab/openhab-addons/files/4860532/summary_report.zip)

Yes it is very strange. In your Report there are issues with the class OJElectronicsBridgeHandler. I renamed the class to OJElectronicsCloudHandler in one of my last commits because i renamed the binding. Also in the files of this pull request there is no file named OJElectronicsBridgeHandler.

I don't understand what's going on her

fwolter commented:
Oops, I didn't run `clean`. Now, it's one warning left. Static fields should be all upper case. You could make it final and move it to the top.

EvilPingu commented:
> Oops, I didn't run `clean`. Now, it's one warning left. Static fields should be all upper case. You could make it final and move it to the top.

All done.

fwolter commented on bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/services/RefreshService.java:
```diff
@@ -0,0 +1,158 @@
+    private @Nullable ScheduledFuture<?> scheduler;
+    private @Nullable Runnable unauthorized;
+    private String sessionId = "";
+    private static Boolean DESTROYED = false;
```
Didn't notice this can't be final. In this case this should be lower case. Sorry.

You could use the primitve type: `boolean`

fwolter commented on bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/OJElectronicsCloudHandler.java:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    private void restartRefreshServiceAsync(long delayInSeconds) {
+        scheduler.schedule(this::ensureSignIn, delayInSeconds, TimeUnit.SECONDS);
```
Can you cancel this task when `dispose()` is invoked?

fwolter commented on bundles/org.openhab.binding.ojelectronics/src/main/resources/ESH-INF/i18n/ojelectronics_de_DE.properties:
```diff
@@ -0,0 +1,31 @@
+# binding
+binding.ojelectronics.name = Binding fï¿½r OJElectronics
```
There seem to be an issue with the encoding. Umlauts take only two bytes in UTF-8, actually.

EvilPingu commented:
@fwolter I corrected the issue. (I hope the enconding fit now)

martinvw commented on bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/models/groups/Schedule.java:
```diff
@@ -0,0 +1,37 @@
+ *
+ * @author Christian Kittel - Initial contribution
+ */
+@NonNullByDefault
```
I think it might be dangerous to annotate your JSON classes like this, it gives a false sense of security. GSON will not validate the `null`-ness of the fields. And I don't know whether your bases it on some documentation but even that is not guaranteed to remain as is.

@cpmeister what do you think of this?

EvilPingu replied:
If i do so, i think i got a warning.

martinvw replied:
> I think it might be dangerous to annotate your JSON classes like this, it gives a false sense of security. GSON will not validate the null-ness of the fields. And I don't know whether your bases it on some documentation but even that is not guaranteed to remain as is.

@openhab/add-ons-maintainers your opinions are greatly appreciated here :-)

martinvw replied:
@EvilPingu if no one veto's this one tonight and you looked into the latest remarks it should be ready to be merged! Thanks for your quick responses.

martinvw commented on bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/models/groups/Schedule.java:
```diff
@@ -0,0 +1,37 @@
+    public List<Day> days = new ArrayList<Day>();
+    @SerializedName("ModifiedDueToVerification")
+    @Expose
+    public Boolean modifiedDueToVerification = false;
```
If you do believe fields will not be `null`, then better use the primitive types.

EvilPingu replied:
I changed this too. I always use primitive types in models now

martinvw commented on bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/models/userprofile/PostSignInQueryModel.java:
```diff
@@ -0,0 +1,102 @@
+     * @param aPIKEY API-Key
+     * @return Model
+     */
+    public PostSignInQueryModel withAPIKEY(String apikey) {
```
See above

EvilPingu replied:
Done

martinvw commented on bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/services/RefreshGroupContentService.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    private void handleThermostat(Thermostat thermostat) {
+        Optional<OJElectronicsThermostatHandler> handler = things.stream()
```
Would something terrible happen if you would support a duplicated device with the same serial number? It would simplify the code:

you could then just skip the `findFirst` and replace it by a `forEach`

EvilPingu replied:
Done

martinvw commented on bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/services/RefreshService.java:
```diff
@@ -0,0 +1,158 @@
+    private final OJElectronicsBridgeConfiguration config;
+    private final Logger logger = LoggerFactory.getLogger(RefreshService.class);
+    private final HttpClient httpClient;
+    private final Gson gson = new GsonBuilder().setPrettyPrinting().setDateFormat("yyyy-MM-dd'T'HH:mm:ss").create();
```
See remark above about using a naming strategy, given the size of the statement you might also consider creating some static method to initialize it.

EvilPingu replied:
I shorten the class names and extracts some methods

martinvw commented on bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/services/RefreshService.java:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    private void refresh() {
+        if (sessionId.equals("")) {
```
It might make more sense to just make this fields nullable, it seems you have "" a magical meaning instead.

If you want to leave, then position the literal first in `String` comparisons - that way if the `String` is `null` you won't get a `NullPointerException`, it'll just return `false`. [source: PMD](http://pmd.sourceforge.net/pmd-4.3.0/rules/design.html#PositionLiteralsFirstInComparisons)

EvilPingu replied:
Done

martinvw commented on bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/services/RefreshService.java:
```diff
@@ -0,0 +1,158 @@
+                GroupContentResponseModel content = gson.fromJson(responseBody, GroupContentResponseModel.class);
+                refreshDone.accept(content);
+            } catch (JsonSyntaxException exception) {
+                logger.error("Error mapping Result to model", exception);
```
In [E.4 of the coding guidelines](http://docs.openhab.org/developers/development/guidelines.html#e-logging) we listed what kind of logging level should be used when. Error should only be used:

> `error` logging should only be used to inform the user that something is tremendously wrong in his setup, the system cannot function normally anymore, and there is a need for immediate action. 

Please change this to use the correct logging levels.

EvilPingu replied:
I think the correct way is to remove the logging and disable the thing. Is this correct?

martinvw replied:
Not really disabling, updating the thing status, calling `updateStatus` and I would see that this indicates a communication error, you could log the full stack trace to debug, but the most important is reporting to the user that something is wrong and passing the `exception.getMessage()` to the `updateStatus` could help for that.

EvilPingu replied:
Done

martinvw commented on bundles/org.openhab.binding.ojelectronics/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,158 @@
+		<config-description>
+			<parameter name="userName" type="text" required="true">
+				<label>User Name</label>
+				<description>User Name for access to OJ Electronics cloud service.</description>
```
I would really not repeat `OJ Electronics` so often, it feels like bad commercial like this :-)

EvilPingu replied:
Yes, I actually use it relatively often.

martinvw commented on bundles/org.openhab.binding.ojelectronics/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,158 @@
+		<label>Regulation Mode</label>
+		<state readOnly="true">
+			<options>
+				<option value="1">Auto</option>
```
Would it make sense to let the binding already map the numeric values to more meaningful strings, it will ease displaying on sitemaps and make rules etc more readable.

EvilPingu replied:
I don't know what you mean exactly. I think this is the mapping of numeric value. Or did you mean i should use string in here and map in the handler in source code?

martinvw replied:
> Or did you mean i should use string in here and map in the handler in source code?

Indeed sorry I was unclear :-)

EvilPingu replied:
Done

EvilPingu commented:
What happens now? Why is this pull request closed? I only changed sign off?

martinvw commented:
> What happens now? Why is this pull request closed? I only changed sign off?

Did you remove the branch before force-pushing? Or did your client do that? GitHub reports it was caused by you 🙂

martinvw commented:
Your branch also became messy, it reports 447 commits of which a lot of them are not yours, did you rebase after merging in the master?

EvilPingu commented:
> Your branch also became messy, it reports 447 commits of which a lot of them are not yours, did you rebase after merging in the master?

I think yes :( I'm so frustrated. I don't know what i have to do now

martinvw commented:
Do you have two remotes? eg. origin and openhab?

EvilPingu commented:
> Do you have two remotes? eg. origin and openhab?

ckit@ubuntu-dev2:~/dev/openhab-2-5-x/git/openhab-addons(2.5.x_evilpingu)$ git remote -v
2.5.x	https://github.com/openhab/openhab-addons.git (fetch)
2.5.x	https://github.com/openhab/openhab-addons.git (push)
EvilPingu	https://github.com/EvilPingu/openhab-addons.git (fetch)
EvilPingu	https://github.com/EvilPingu/openhab-addons.git (push)
origin	https://github.com/openhab/openhab-addons.git (fetch)
origin	https://github.com/EvilPingu/openhab-addons.git (push)

martinvw commented:
What I would propose:

Create backup:

```
git branch backup
git push -u EvilPingu backup
```

Soft reset maintaining all changes but discarding all commits:

```
git fetch 2.5.x
git reset 2.5.x/2.5.x
```

Redoing it all in a single commit 

```
git add bundles/org.openhab.binding.ojelectronics/
git add CODEOWNERS bom/openhab-addons/pom.xml bundles/pom.xml
git commit -ms "[ojelectronics] Initial contribution"
```

And if the local commit looks fine you can force push it

Please also check out the status of bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/OJElectronicsCloudHandler.java~HEAD

PS: I would rename the remote 2.5.x because it might give problems with branch name vs remote name

EvilPingu commented:
> What I would propose:
> 
> Create backup:
> 
> ```
> git branch backup
> git push -u EvilPingu backup
> ```
> 
> Soft reset maintaining all changes but discarding all commits:
> 
> ```
> git fetch 2.5.x
> git reset 2.5.x/2.5.x
> ```
> 
> Redoing it all in a single commit
> 
> ```
> git add bundles/org.openhab.binding.ojelectronics/
> git add CODEOWNERS bom/openhab-addons/pom.xml bundles/pom.xml
> git commit -ms "[ojelectronics] Initial contribution"
> ```
> 
> And if the local commit looks fine you can force push it (maybe force validate the existence of the backup, but otherwise I still have your branch here :-0)
> 
> Please also check out the status of bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/OJElectronicsCloudHandler.java~HEAD
> 
> PS: I would rename the remote 2.5.x because it might give problems with branch name vs remote name

I don't exactly know what happens now: But i did it :D Many many thanks

EvilPingu commented on bundles/org.openhab.binding.ojelectronics/README.md:
```diff
@@ -0,0 +1,64 @@
+
+At this moment all information read only.
+
+## Road map
```
Done

EvilPingu commented on bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/OJElectronicsThermostatHandler.java:
```diff
@@ -0,0 +1,110 @@
+@NonNullByDefault
+public class OJElectronicsThermostatHandler extends BaseThingHandler {
+
+    private String serialNumber;
```
That's true.

EvilPingu commented on bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/OJElectronicsThermostatHandler.java:
```diff
@@ -0,0 +1,110 @@
+ * The {@link OJElectronicsThermostatHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author EvilPingu - Initial contribution
```
Done

EvilPingu commented on bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/OJElectronicsThermostatHandler.java:
```diff
@@ -0,0 +1,110 @@
+     */
+    public OJElectronicsThermostatHandler(Thing thing) {
+        super(thing);
+        serialNumber = getConfigAs(OJElectronicsThermostatConfiguration.class).serialNumber;
```
I think yes because it is required in thing-types.xml

EvilPingu commented on bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/config/OJElectronicsBridgeConfiguration.java:
```diff
@@ -0,0 +1,59 @@
+    /**
+     * Software Version
+     */
+    public Integer softwareVersion = 1060;
```
Done

EvilPingu commented on bundles/org.openhab.binding.ojelectronics/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,158 @@
+				<description>Password for access to OJ Electronics cloud service.</description>
+				<context>password</context>
+			</parameter>
+			<parameter name="apiKey" type="text" required="true" min="36" max="36">
```
I changed this in the documentation.

EvilPingu commented on bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/models/userprofile/PostSignInQueryModel.java:
```diff
@@ -0,0 +1,102 @@
+
+    @SerializedName("APIKEY")
+    @Expose
+    public String apikey = "";
```
Done

EvilPingu commented on bundles/org.openhab.binding.ojelectronics/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,158 @@
+		<supported-bridge-type-refs>
+			<bridge-type-ref id="ojcloud"/>
+		</supported-bridge-type-refs>
+		<label>OJ Electronics OWD5/MWD5 Thermostat</label>
```
Done

EvilPingu commented on bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/models/groups/Day.java:
```diff
@@ -0,0 +1,37 @@
+public class Day {
+
+    @SerializedName("WeekDayGrpNo")
+    @Expose
```
Of course it works without it. I generated this models with http://www.jsonschema2pojo.org/. I remove this explicit expose. No problem

EvilPingu commented on bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/models/groups/Day.java:
```diff
@@ -0,0 +1,37 @@
+@NonNullByDefault
+public class Day {
+
+    @SerializedName("WeekDayGrpNo")
```
I changed this in all models

EvilPingu commented on bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/OJElectronicsCloudHandler.java:
```diff
@@ -0,0 +1,158 @@
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+    }
```
Done.

martinvw commented on bundles/org.openhab.binding.ojelectronics/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -121,13 +121,13 @@
-				<option value="Frost Protection">Frost Protection</option>
-				<option value="Boost">Boost</option>
-				<option value="Eco">Eco</option>
+				<option value="auto">Auto</option>
```
It seems you map to the capitalized values, or am I mistaken?

EvilPingu replied:
You are right. I have to use lower case letters here and in source code too.

martinvw commented on bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/ThermostatHandler.java:
```diff
@@ -0,0 +1,193 @@
+        return REGULATION_MODES.get(regulationMode);
+    }
+
+    private static HashMap<Integer, String> createRegulationMap() {
```
Declare vars with their interface for the `Collection` types, so this should rather be: 
```suggestion
    private static Map<Integer, String> createRegulationMap() {
```

martinvw commented on bundles/org.openhab.binding.ojelectronics/src/main/java/org/openhab/binding/ojelectronics/internal/ThermostatHandler.java:
```diff
@@ -0,0 +1,193 @@
+    }
+
+    private static HashMap<Integer, String> createRegulationMap() {
+        HashMap<Integer, String> map = new HashMap<Integer, String>();
```
Please use the diamond operator, ie replace the second occurrence of the same generic by `<>`

```suggestion
        Map<Integer, String> map = new HashMap<>();
```

EvilPingu replied:
Done.

EvilPingu commented on bundles/org.openhab.binding.ojelectronics/README.md:
```diff
@@ -0,0 +1,64 @@
+| comfortEndTime     | Date time          | Date and time when the thermostat switchs back from comfort mode to automatic mode |
+| boostEndTime       | Date time          | Date and time when the thermostat switchs back from boost mode to automatic mode   |
+| manualModeSetpoint | Number:Temperature | Target temperature of the manual mode                                              |
+| vacationEnabled    | Switch             | Vacation is enabled                                                                |
```
Writing Documentation: My Favorite Occupation. :grin:

martinvw replied:
Thanks for reminding me I would forget about this one 😛

martinvw replied:
The visual studio code plugin can help you generate a least the items definition, I would not know by heart whether I can also generate the others.

EvilPingu replied:
> Thanks for reminding me I would forget about this one stuck_out_tongue

Damn. That would have been my chance ... :rofl: 

But it's no problem... I wrote a good example that anyone can start with. At least that's what I hope.

martinvw commented:
Thanks a lot!

EvilPingu commented:
It was a great pleasure for me. I learned a lot. :+1:
